### PR TITLE
[Magiclysm] Add bag of infite knives, allow to specify more conditions for artifact recharging

### PR DESCRIPTION
**Pull request by Salty-Panda**
#### Summary
Mods "[Magiclysm] Add bag of infinite knives"

#### Purpose of change

I wanted it.

#### Describe the solution

Added recharge_condition::worn, wielded, held to artifacts. It works like periodic, but require the item to be worn/wielded
Added the bag, it casts a spell that summons one fake item, a throwing knife. The knife is kinda lackluster, in the future the enchantments will be able to modify skills though, so it shall be updated.
Changed description and reduced the price ( KorGenT blocked wearing multiple rings at once recently ) of an existing ring that was doing something similar.

#### Describe alternatives you've considered

Adding the possibility to modify skills via enchantment, but KorGenT already has specific plans on how to do it.

Adding a greater version that randomly summons variant dealing electric/pure/more piercing/acid damage. would require cpp support too.

#### Testing

Spawned, used

### Additional context
Active artifacts that cast spells and reload on their own could be abused by wearing a few "Ring of cast strong spell" in a backpack and just equipping them when needed in place of a staple + STAT.
Forcing PC to wear items in order to recharge them somewhat reduces the problem.
Also, that ring was outrageously expensive considering the uselessness of throwing knives. It could only summon a few, required an hour to recharge a single-use, and was half again as expensive as a ring with a permanent 1 STR bonus.

### CODE REVIEW

**Review by KorGgenT — CHANGES_REQUESTED** [review]
in the artifact code, worn should not be a "charge type" it should be a "prerequisite"; that way you can have worn be required for any charge type, isntead of adding worn enums forr each charge type. see also: how i implemented artifact::has

**Review by KorGgenT — CHANGES_REQUESTED** [review]
in the artifact code, worn should not be a "charge type" it should be a "prerequisite"; that way you can have worn be required for any charge type, isntead of adding worn enums forr each charge type. see also: how i implemented artifact::has

**Comment by Salty-Panda** [comment]
> in the artifact code, worn should not be a "charge type" it should be a "prerequisite"; that way you can have worn be required for any charge type, instead of adding worn enums for each charge type. see also: how i implemented artifact::has

So I should add a struct to relic.h
```cpp
enum relic_recharge_condition {
    solar_sunny,
    worn,
    num
};
```
and to recharging function some check before the switch with `charge.type`?
```cpp
    if(charge.requirement == relic_recharge_condition::worn){
        if( ! ( item.is_wield || item.is_held ) ) ) {
            return;
        }
    }
```

Essentially make it work in a way similar to how "has" and "condition" behave?

**Comment by KorGgenT** [comment]
> So I should add a struct to relic.h [...]
yes, exactly. sorry it took me so long to get back to your question!

**Comment by Salty-Panda** [comment]
@KorGgenT done and tested

**Comment by Salty-Panda** [comment]
Forgot to add spawning for POC item though, will add to some item groups today

**Comment by Salty-Panda** [comment]
Added a few variants using WONDER flag, sadly you can't use copy-from in spells.
Increased price since it's more powerful overall and added it to itemgroups

File: src/relic.cpp [hunk 1]
```diff
@@ -429,29 +446,65 @@ void relic::try_recharge( item &parent, Character *carrier, const tripoint &pos
...
             return;
         }
-        case relic_recharge::periodic: {
+        case relic_recharge_type::PERIODIC: {
             charge.accumulate_charge( parent );
             return;
         }
-        case relic_recharge::solar_sunny: {
+        case relic_recharge_type::SOLAR_SUNNY: {
+            debugmsg( "solar" );
```
Comment (Salty-Panda) [inline src/relic.cpp:462]:
```suggestion
```

**Comment by Salty-Panda** [comment]
The only failed test is vehicle_turrets_test.cpp
AFAIK, this PR is finally done and ready to review

**Comment by KorGgenT** [comment]
i don't have the time to review this tonight, but have you tested how loading previously saved artifacts will work with your change? it's important that they be able to load without errors.

**Comment by Salty-Panda** [comment]
Loaded save with two artifacts, Dim Sky(2/2) and Fire Rain(1/1), used both and both refilled to full.
No errors, no visible debug messages

**Pull request merged**